Channel refactor

canndrew opened the pull request:
Refactor DNL's `Channel` type in order to (a) separate state which is saved to the wallet from transient state which is discarded on disconnection, (b) simplify the API to make it less confusing, less bug-prone and require less duplicated effort between DNL and geewallet.

canndrew commented:
This is what was posted on the telegram channel w.r.t. getting these changes into a mergable state:

> Andrés G. A., [11.02.21 15:39]
> bb70f990ed0956cf88b7f21badc3039964080c8a -> ok, but remove code (don't leave commented code)
> 
> Andrés G. A., [11.02.21 15:39]
> 6299085a4eb9431d7a23356f097cfdc1d2178b18 -> you already did this one for coldstorage support right?
> 
> Andrés G. A., [11.02.21 15:40]
> facad529dd3cdfd9c7fde2b94a42872cd15faa6d -> ok if it's commented code in DNL upstream (if it's commented code introduced by the previous commits, let's remove it in those commits)
> 
> Andrés G. A., [11.02.21 15:41]
> 74c33c9d4d693b2eb61b9c75e207ca8d8a0761c3 -> ok, , but remove code (don't leave commented code)
> 
> Andrés G. A., [11.02.21 15:41]
> c0820b857c4d638c059f4f25bc77ef0426844bd9 -> ok, , but remove code (don't leave commented code)
> 
> Andrés G. A., [11.02.21 15:42]
> 1abfd71dc52e04f031b3b8d27af0d46178ed2f05 -> ok
> 
> Andrés G. A., [11.02.21 15:42]
> c41f7cb4b1c1cb033f21c33a0aa7eeb29a1658ca -> ok
> 
> Andrés G. A., [11.02.21 15:43]
> 520e385ff182e97c121ec4cd44050bf782f86929 -> ok
> 
> Andrés G. A., [11.02.21 15:43]
> 498786d6592753bef6face3a5e37ca19ee24a6d7 -> ok
> 
> Andrés G. A., [11.02.21 15:44]
> 7f6723fb87c5736e04c87d3cc8cfbf75ffbba004 -> only remove unused type, don't change "readability" aspects
> 
> Andrés G. A., [11.02.21 15:45]
> [In reply to Andrés G. A.]
> can do in a lower-priority PR later
> 
> Andrés G. A., [11.02.21 15:46]
> 000a3d8036c61779d4270080037e8fce5dab2eef -> ok
> 
> Andrés G. A., [11.02.21 15:46]
> b8e82719a41f26d08ca810b88fe5dfa118cdecfc -> ok
> 
> Andrés G. A., [11.02.21 15:48]
> 78c166cc9c9ceeeee5bdd9b8cd9ff388bea29bf2 , 54da3756894d6a0c072f06cd7325635f63ea7acc , f5af1f886cad0d3ad7b40a0ad6a3c632247d8b5f , 4dc3c7f668bdcb5b59193b6a3454672255c45522 -> please squash these 4 commits into 1
> 
> Andrés G. A., [11.02.21 15:49]
> a24f64b41cfdb6664c2b00ddc6e78d34ca1f856b, f8f035ed073a40acd9535e156d26a92bc4df3000 -> please squash these 2 commits into 1
> 
> Andrés G. A., [11.02.21 15:50]
> 375f9dcc190d025070c1ab14b586369110bed5f5 -> ok
> 
> Andrés G. A., [11.02.21 15:51]
> 5333bbe45e8d196f5f428281b4dbe43015f1923c -> ok
> 
> Andrés G. A., [11.02.21 15:52]
> b0778cae17f43f0f1748dc95db874de0c4736aca -> ok
> 
> Andrés G. A., [11.02.21 15:52]
> e4d6816afedcb788014932566729a43b02aa00cd -> ok
> 
> Andrés G. A., [11.02.21 15:55]
> [In reply to Andrés G. A.]
> 0cb47ea273647b3a736d823adf125531c438af91 -> squash with the 2 above
> 
> Andrés G. A., [11.02.21 15:57]
> 769318bcac85c50ccb4f1aa41253b3589672e292, c02298bbdaaa1a394101e6efbad6c723e240f027, 9a2cdb4c3ecda893331a4c4ee36223136f0c1bfe -> squash these 3 into 1
> 
> Andrés G. A., [11.02.21 15:59]
> ea58de89c0f492ed64b79d79dedffbc5aaa92d81 -> ok
> 
> Andrés G. A., [11.02.21 16:00]
> 680a20e1d3a054299c23f0c1967c7cb9c0c1f972 -> ok
> 
> Andrés G. A., [11.02.21 16:00]
> 296e1f1aadbee5ab006539c8773d83d1052ae0e3 -> ok
> 
> Andrés G. A., [11.02.21 16:00]
> 1af4b7446ee732ce6b8db7909d1a847b7a08a1b7 -> ok
> 
> Andrés G. A., [11.02.21 16:01]
> 7dd1f0036f4c71f5cfcc60ffb502d88d693160de -> ok
> 
> Andrés G. A., [11.02.21 16:01]
> 51bcf5c74a5a31ffd7b1473748c3b68646529efa -> ok
> 
> Andrés G. A., [11.02.21 16:02]
> 561257e8bff17814e4244a03d2fea0fe4729d55a -> ok
> 
> Andrés G. A., [11.02.21 16:07]
> 661d6744744d8e09b1160c1b39f5d39fc35f30d2 -> ok
> 
> Andrés G. A., [11.02.21 16:07]
> 27d2d49cf6efcfe2c3583ef8c8ab0d35293fe9a6 -> ok
> 
> Andrés G. A., [11.02.21 16:07]
> 130cfd66fcc9174c509c8f7bb76f1f61ab29ca2c -> ok
> 
> Andrés G. A., [11.02.21 16:08]
> 8ee9cdc4e32d1e2be7039bd3a23a63bffe76f8b9 -> ok
> 
> Andrés G. A., [11.02.21 16:10]
> f5036a70c26982a2ed41852074ebbfc49f73eca4 -> ok
> 
> Andrés G. A., [11.02.21 16:13]
> 8c490b8ab69b59c11007e7e5b8795ba4820813ec -> ok
> 
> Andrés G. A., [11.02.21 16:13]
> 6e6daefd3fd42e5504af94e194faefc914681b93 -> ok
> 
> Andrés G. A., [11.02.21 16:15]
> c449e7e7aec9ace6ff72a7fe0e26c5f8c681c027 -> ok but don't rename members to have Opt suffix
> 
> Andrés G. A., [11.02.21 16:17]
> eecbba81c742bda776715a52c60f00c04abd7a34 -> I don't think this is needed; I understand the need to use Opt suffix for variables (because you have fooOpt and inside the match pattern you then later have real `foo`) but there's no need to do it on members, otherwise whenever they change their type they need to be renamed
> 
> Andrés G. A., [11.02.21 16:24]
> 1fb3bf70d58021bf20a75f1d2c315e5851f6ae1e -> ok
> 
> Andrés G. A., [11.02.21 16:25]
> c4bfaf549c1463c33c51eb483f1db847c3489f0f -> ok
> 
> Andrés G. A., [11.02.21 16:26]
> 8d5107edcce16d96406e1c3fc35adbf246383323 -> ok
> 
> Andrés G. A., [11.02.21 16:26]
> 7a2571e4d722e6341846b00f2315e2b58ff701c8 -> ok
> 
> Andrés G. A., [11.02.21 16:27]
> 51e8c49eaadc153b735753c7697560833b4fa455 -> ok
> 
> Andrés G. A., [11.02.21 16:31]
> aaa0aa1d66bcc6d84763580da6dd0fe61dd91829, ef8e51d06f3fe2bae2c24569c7022cd3d14cc335, 136aa4ea2403edc2e8ae04a2a99506d684e69f6a, e5b761f9a83e83b60526f14f6442cd2b5bb55cf5, 15bad14c1a409794d6476d86e649c4bf663360e4, 9d2177575799165e9739653de3eeeafed431a770, 04347881b359c8bf49e99eb78ad98027d99af875, 5c5ef448fa3b2231a5cd89e3dd6328175ce7e2db, f4d283d16f8456cf0e4a08783a61efa8ceaed219 -> squash all these removals into 1 pls, as they all seem related
> 
> Andrés G. A., [11.02.21 16:31]
> 7b08bd8338409b38ca282c7b8c2198a3faabdaed -> ok
> 
> Andrés G. A., [11.02.21 16:32]
> 51ba93f678abf1d30ab7184fd091ca9005b2a4e1 -> ok
> 
> Andrés G. A., [11.02.21 16:32]
> 355f8630dec0946196a0f3831403c4ee6a799b8a -> ok
> 
> Andrés G. A., [11.02.21 16:34]
> 53549777e16eed4e54aa7363ecdea3f4f80b26d9 -> ok
> 
> Andrés G. A., [11.02.21 16:36]
> 7e9c3621c035e40c6d79d8bb4744a2d9c85d2de4 -> ok, but don't add commented code (just remove the code), and remove the Opt suffix
> 
> Andrés G. A., [11.02.21 16:37]
> 4582279d3288e3476ba519e55049fe585d03b53f -> ok, but remove the Opt suffix
> 
> Andrés G. A., [11.02.21 16:38]
> 70ac292f2d33df6c954993132d924b04ff1afe50 -> if it can be derived, then leave the TxId as a getter-only property?
> 
> Andrés G. A., [11.02.21 16:39]
> 3658a3b30abb95fee20e250c766201bb4662b32b -> ok
> 
> Andrés G. A., [11.02.21 16:39]
> 2ef038b0cbd4c7ee43d12fb711e9bb67e1a27328 -> ok
> 
> Andrés G. A., [11.02.21 16:40]
> f2d50e9de91206b65d825ea83f8aa7c2041bcc5d -> ok
> 
> Andrés G. A., [11.02.21 16:40]
> 2935f400350f616d2ae472f943e3324f9472dd53 ->ok

knocte commented:
Please don't forget that this needs to be rebased against Joe's master, not DNL.Kiss' master. But don't propose PR there until I have re-reviewed myself.

canndrew commented:
I've made the requested changes with a couple of exceptions:

  *  0cb47ea -> squash with the 2 above

This didn't make a lot of sense to me. I'm not sure why it should be squashed into the two before it.

  *  7f6723f -> only remove unused type, don't change "readability" aspects

Maybe I shouldn't have change the formatting of `ChannelTypes.fs` in the first place, but reverting the formatting would probably take a couple of hours since there's 50 or so commits after it which change that file. Just removing the `Opt` suffixes was a lot of work and that's a smaller change that had to be rebased through far fewer commits. Also, at the end of the rebase, that file will get stripped down to almost nothing anyway. Also the new formatting is lot more readable (for me at least). I'd prefer to just not do this unless you feel very strongly about it. Going forward I won't make these kinds of changes for no good reason.

I've also started rebasing on top of joe's commits. So this now almost up to date with joe's master, with the exception of the most recent commit (`FundingTxProvider` removal) since that'll be a bit of work to rebase on. I haven't rebased-away the DNL.Kiss changes yet either, except for the "Prevent possible F#4.0 issue" commit (ie. `ResultUtils.Portability`) since that's included in joe's master.

knocte commented:
> This didn't make a lot of sense to me. I'm not sure why it should be squashed into the two before it.

Fair enough, but you can merge the 2 above into 1? (merge 2 into 1 instead of 3 into 1).

> but reverting the formatting would probably take a couple of hours since there's 50 or so commits after it which change that file

😬 

> Going forward I won't make these kinds of changes for no good reason.

But the real way to learn this properly is feeling the pain 😅 .
Well let's reach a compromise: you can leave the formatting changes there, so long as you split the commit into 2 (one that just does formatting changes, and the other that removes the unused type).

knocte commented:
> so long as you split the commit into 2

FYI this should only take 5 minutes if you use `git rebase -i` and the `edit` keyword.

canndrew commented:
> Fair enough, but you can merge the 2 above into 1? (merge 2 into 1 instead of 3 into 1).

Done.

> split the commit into 2 (one that just does formatting changes, and the other that removes the unused type).

Done. That sort of thing doesn't take long :)

canndrew commented:
I've rebased on top of the joe's master now. So we have the `FundingTxProvider`-removal commit.

I've also fixed the warnings on joe's master and I'll rebase on top of that next and make sure none of my commits introduce any new warnings and that tests still pass after each one (this can be automated with `git rebase -i`, assuming everything goes well).

canndrew commented:
I've rebased to fix all warnings so that every commit builds cleanly. I've then started replacing the channel operations/events system with simple methods (just `CreateChannelReestablish` done so far). This will make it easier to take the next step of refactoring the `Channel` and `Commitments` types into two new types - one which stores the channel state which is cryptographically committed to and saved to the wallet and one which stores transient channel state which only exists for the lifetime of the TCP connection and gets reset on disconnect. Then we should be able to reestablish channels correctly.

canndrew commented:
I've added a bunch more commits here. I've removed the `ChannelOperation`/`ChannelEvent` system and replaced all the channel operations with methods.

For an overview of what this means: Previously DNL's `Channel` type had an `executeCommand` method which took a `ChannelOperation` value specifying the operation you want to perform. eg. It looked like this (simplified):

```fsharp
type ChannelOperation =
    | UpdateFee of newFeeRate: FeeRate
    
    // more operations here


let executeCommand (channel: Channel)
                   (operation: ChannelOperation)
                       : ChannelEvent
    match operation with
    | UpdateFee of newFeeRate ->
        // do the fee update
    
    // handle the other operations here
```

When you called `executeCommand` you'd have to look at the `ChannelEvent` it returned in order to get the result of executing the command. For instance, there's a `ChannelEvent` called `WeAcceptedOperationUpdateFee` which gets returned when you execute the `UpdateFee` command and which gives you the `update_fee` message that you have to send.

```fsharp
type ChannelEvent =
    | WeAcceptedOperationUpdateFee of UpdateFeeMsg
```

So you do a fee update like this (simplified):

```fsharp
let operation = UpdateFee newFeeRate
let channelEvent = Channel.executeCommand channel operation
let updateFeeMsgToSend =
    match channelEvent with
    | WeAcceptedOperationUpdateFee updateFeeMsgToSend -> updateFeeMsgToSend
    | _ -> failwith "unexpected event"
send updateFeeMsg
```

Except you then also have to pass the event back into the `Channel` in order to update the channel's state, in this case so that the channel knows what the new fee rate is. You do this by calling `applyEvent` which you have to call every time you call `executeCommand` and give it back the data it just gave you. So the full(er) fee updating code actually looks like this:

```fsharp
let operation = UpdateFee newFeeRate
let channelEvent = Channel.executeCommand channel operation
let updateFeeMsgToSend =
    match channelEvent with
    | WeAcceptedOperationUpdateFee updateFeeMsgToSend -> updateFeeMsgToSend
    | _ -> failwith "unexpected event"
send updateFeeMsg
let newChannel = Channel.applyEvent channel channelEvent
```

Most of these events also include fields which are internal implementation details of DNL and which are only there so that DNL can use them to update it's state when you pass the event back to it with `applyEvent`.

I've streamlined this so that every `ChannelOperation` is now just a method on `Channel`. eg. There's now a method:

```fsharp
type Channel with
    member self.UpdateFee (newFeeRate: FeeRate): Channel * UpdateChannelMsg
```

This will perform the state-update internally and return the new channel object along with the user-relevant data from the event, in this case the `update_fee` message you need to send.

This makes the API a lot saner, simpler and safer. In terms of refactoring this is useful because before there was a split where some logic existed in `executeCommand` and some in `applyEvent`. Concatenating the code for each operation+event pair into it's own method makes the behaviour of the code easier to see (for me at least). It also revealed a few places where code was just redundant - eg. data was being created by `executeCommand` and then thrown away again in `applyEvent`, despite the data not being relevant to the caller either. So stripping that stuff out will make the next refactoring steps easier. This also lets me think about exactly where these functions will need to go when I split the `Channel` type into two levels - a cryptographically-signed, wallet-serialized level and a transient, current-connection-relevant level.

canndrew commented:
I've pushed a few small commits which just remove unused cruft. I'm in the middle of refactoring the commitment-handling code and in particular the code which adds/removes htlcs to/from commitment specs. I don't want to spend too long dealing with HTLC stuff (since we don't yet support HTLCs in geewallet) and I'm trying to do this without changing any behaviour, however as far as I can tell the code I'm looking at is already completely broken, which makes it kinda hard.

canndrew commented:
I've pushed a commit which fixes a problem with the way HTLCs were being handled. I'm working on some larger changes which change how commitment updates (ie. htlc-updates, fee updates and monohop-payment updates) are conglomerated in to commitment spec. This will allow us to separate the saved/signed channel commitment states which go into the wallet from the current in-progress commitment state which hasn't been committed to.
